File: Session_05_DynamicProgramming_Basics/multi_language_implementations/coin_change.swift
import Foundation

let INF = 1000000000

func getMinCoins(_ coins: [Int], _ amount: Int) -> Int {
    var dp = [Int](repeating: INF, count: amount + 1)
    dp[0] = 0
    for i in 1...amount {
        for j in 0..<coins.count {
            if coins[j] <= i {
                let subRes = dp[i - coins[j]]
                if subRes != INF && subRes + 1 < dp[i] {
                    dp[i] = subRes + 1
                }
            }
        }
    }
    let result = dp[amount]
    return result == INF ? -1 : result
}

func main() {
    let coins = [1, 3, 4]
    let amount = 6
    print("Min coins required: \(getMinCoins(coins, amount))")
    print("\nPress Enter to exit...", terminator: "")
    fflush(stdout)
    let _ = readLine()
}

main()

File: Session_05_DynamicProgramming_Basics/multi_language_implementations/coin_change.js
const readline = require('readline');

const INF = 1e9;

function getMinCoins(coins, amount) {
    const dp = new Array(amount + 1).fill(INF);
    dp[0] = 0;
    for (let i = 1; i <= amount; i++) {
        for (let j = 0; j < coins.length; j++) {
            if (coins[j] <= i) {
                const subRes = dp[i - coins[j]];
                if (subRes !== INF && subRes + 1 < dp[i]) {
                    dp[i] = subRes + 1;
                }
            }
        }
    }
    const result = dp[amount];
    return result === INF ? -1 : result;
}

function main() {
    const coins = [1, 3, 4];
    const amount = 6;
    console.log(`Min coins required: ${getMinCoins(coins, amount)}`);
    console.log();

    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });
    rl.question("Press Enter to exit...", () => {
        rl.close();
    });
}

main();
